app/src/org/commcare/dalvik/activities/PinAuthenticationActivity.java

```diff
@@ -0,0 +1,173 @@
+package org.commcare.dalvik.activities;
+
+import android.content.Intent;
+import android.os.Bundle;
+import android.util.Log;
+import android.view.View;
+import android.widget.Button;
+import android.widget.EditText;
+import android.widget.TextView;
+import android.widget.Toast;
+
+import org.commcare.android.database.app.models.UserKeyRecord;
+import org.commcare.android.framework.ManagedUi;
+import org.commcare.android.framework.SessionAwareCommCareActivity;
+import org.commcare.android.framework.UiElement;
+import org.commcare.dalvik.R;
+import org.commcare.dalvik.application.CommCareApplication;
+import org.javarosa.core.services.locale.Localization;
+
+/**
+ * When a user selects Set/Reset PIN from the home activity options menu, they must first
```

Comment:
I wonder whether we truly need the capacity to check someone's PIN/PW after they've logged in. I know it's super common, but it's kind of awkward if you just logged someone in to change their PIN and you need to enter the PW twice, right? Maybe good to double check with product (fine to leave for now until we can UAT)